doc/bulk_counter/bulk_counter.md

```diff
@@ -0,0 +1,145 @@
+
+### Requirements
+
+- Syncd shall use bulk stats APIs based on object type. E.g. for a counter group that queries queue and pg stats, queue stats support bulk while pg stats does not, in that case queue stats shall use bulk API, pg stats shall use non bulk API
+- For a certain object type in a counter group, it shall use bulk stats only if:
+  - The stats capability for each counter IDs shall include SAI_STATS_MODE_BULK_READ, SAI_STATS_MODE_BULK_CLEAR, SAI_STATS_MODE_BULK_READ_AND_CLEAR (avoid handling stats mode change)
+  - Each object queries exactly the same counter IDs. (Requirement from function signature of sai_bulk_object_get_stats and sai_bulk_object_clear_stats)
+- Syncd shall automatically fall back to old way if bulk stats APIs are not supported
+- Syncd shall utilize API sai_query_stats_capability to query bulk capability. Syncd shall treat counter as no bulk capability if API sai_query_stats_capability return error except SAI_STATUS_BUFFER_OVERFLOW (SAI_STATUS_BUFFER_OVERFLOW requires a retry with larger buffer)
+- Syncd shall call bulk stats API in flex counter thread and avoid calling it in main thread to make sure main thread only handles short and high priority tasks. (This is the default behavior in flex counter infrastructure)
+- In phase 1, the change is limited to syncd only, no CLI/swss change. Syncd shall use bulk stats API by querying the statistic capability. Syncd shall deduce the bulk stats mode according to the stats mode defined in FLEX DB:
+  - SAI_STATS_MODE_READ -> SAI_STATS_MODE_BULK_READ
+  - SAI_STATS_MODE_READ_AND_CLEAR -> SAI_STATS_MODE_BULK_READ_AND_CLEAR
+
+### Architecture Design
+
+For each counter group, different statistic type is allowed to chooose bulk or non-bulk API based on SAI capability.
+
+![architecture](/doc/bulk_counter/bulk_counter.svg).
+
+> Note: In the picture, pg/queue watermark statistic use bulk API and buffer watermark statistic uses non-bulk API. This is just an example to show the design idea. 
+
+### High-Level Design
+
+Changes shall be made to sonic-sairedis to support this feature. No CLI change. No DB schema change.
+
+> Note: Code present in this design document is only for demonstrating the design idea, it is not production code.
+
+#### sonic-sairedis
+
+##### Bulk Statistic Context
+
+A new structure shall be added to FlexCounter class. 
+
+This structure is created because:
+
+- Meet the signature of sai_bulk_object_get_stats and sai_bulk_object_clear_stats
+- Avoid constructing these information each time collecting statistic. The bulk context shall only be updated under below cases:
+  - New object join counter group. E.g. adding a new port object.
+  - Existing object leave counter group. E.g removing an existing port object.
+
+```cpp
+struct BulkStatsContext
+{
+    bool support = true;
+    std::vector<sai_object_id_t> object_vids;
```

Comment:
Field `object_vids` looks needless, since having `object_keys`. In addition, `object_type` should be added for self-awareness.